### PR TITLE
Added the ability to download cached images.

### PR DESCRIPTION
**Pull request by MrAntonS**
Please take a look at this code. It works on Android but I don't see any permission required popup, yet I can download images.

### CODE REVIEW

**Comment by hjiangsu** [comment]
I took a quick look at this, and theres a couple of things to note potentially:

From what I saw, the `image_gallery_saver` requires the following permissions:
- `NSPhotoLibraryAddUsageDescription` for iOS
- `<application android:requestLegacyExternalStorage="true" .....>` for Android 10 versions, and maybe some other storage permissions for other Android versions

I'm not too sure if we have to directly specify these permissions in the AndroidManifest.xml and Info.plist files, but that might  be the case

What I can do for now is to merge this PR into the `develop` branch so that more tests can be done on a release build for Android and iOS before I merge it into the main branch

Do you have access to a physical Android device to install the release apk on? If you do, then you can do the following:
- Theres a script to build the release versions for both android and ios. Just run `dart scripts/build.dart` in the terminal and it'll handle the rest (you might have to comment out the part that builds the iOS IPA file if you're not on a Mac)
- The release files will be stored in `/releases` directory, and you should be able to then install that APK file on your phone
- Try to do it with a clean install if possible, so delete the existing app and install the build from the `releases` directory

Thanks again!

**Comment by MrAntonS** [comment]
Yes i have an android and I'm testing release apks. It worked without asking enything, but I added the permission to android manifest. I want this permission pop-up to appear. I'll do some more research on this on Monday(no access right now) 😅

**Comment by hjiangsu** [comment]
> It worked without asking enything, but I added the permission to android manifest. I want this permission pop-up to appear. I'll do some more research on this on Monday(no access right now) 😅

Sounds good, thanks for all your work so far!

**Comment by MrAntonS** [comment]
Not at all, It's in the implementation phase anyway

**Comment by MrAntonS** [comment]
> This is what I found: https://pub.dev/packages/permission_handler
> 
> Seems straightforward enough.

Wonder why the include Windows in platform if it's obviously not supported XD. Anyway, looks very simple. Will test this one.

**Comment by MrAntonS** [comment]
Alright, one last thing i want to add is to change download icon to check-mark once saved. I'll do that tomorrow. Please check if the download works fine on IOS and add any code if necessary

**Comment by hjiangsu** [comment]
Hey! Thanks again - I'll go ahead and check it on iOS

I'll also do a quick look through and pass back any questions I have about the implementation :D

File: android/app/build.gradle [hunk 1]
```diff
@@ -24,7 +24,7 @@ if (flutterVersionName == null) {
 
 android {
     namespace "com.hjiangsu.thunder"
-    compileSdkVersion flutter.compileSdkVersion
+    compileSdkVersion 33
```
Comment (hjiangsu) [inline android/app/build.gradle:27]: Is this change necessary? Just wanted to know if there was a specific reason for this!

File: lib/shared/image_viewer.dart [hunk 2]
```diff
@@ -1,61 +1,75 @@
...
-                    inPageView: false,
-                    initialAlignment: InitialAlignment.center,
-                    reverseMousePointerScrollDirection: true,
-                    gestureDetailsIsChanged: (GestureDetails? details) {},
-                  );
+      body: SafeArea(
+        child: Stack(
+          alignment: Alignment.topRight,
+          children: [
+            PhotoView(
```
Comment (hjiangsu) [inline lib/shared/image_viewer.dart:34]: hmm, this was changed by me a couple days ago. In image_viewer, I'm no longer using the `PhotoView` widget - but rather `ExtendedImageSlidePage`

Can you double check to see if maybe there was a commit which forced this change back to what it was previously?

File: pubspec.yaml [hunk 3]
```diff
@@ -61,7 +61,12 @@ dependencies:
   webview_flutter_wkwebview: ^3.5.0
   share_plus: ^7.0.2
   flex_color_scheme: ^7.1.2
+  image_gallery_saver: ^2.0.2
+  path_provider: ^2.0.15
+  flutter_cache_manager: ^3.3.0
+  permission_handler: ^10.3.0
   dynamic_color: ^1.6.5
+  baseflow_plugin_template: ^2.1.2
```
Comment (hjiangsu) [inline pubspec.yaml:69]: Just double checking, but what is `baseflow_plugin_template` used for?

**Review by hjiangsu — COMMENTED** [review]
I wrote up some comments from the changes - one main thing that I noticed was that the `image_viewer.dart` changes reverted back some of the stuff I did a couple days ago.

Not sure if this was intentional, but I would like to have it changed back! The changes that I made to `image_viewer` was pretty much improvements to the slide down to dismiss gestures to make it smoother

**Comment by guigs4** [comment]
> Wonder why the include Windows in platform if it's obviously not supported XD.

They do have a separate lib for windows that depends on the mobile one. Maybe it all would be handled by the main one at some point and someone forgot to update the tag once it was changed.

But that's all speculation. 🤔

**Comment by MrAntonS** [comment]
Here. One thing, i fetched updated from main, so there might be some files change for develop.

**Comment by hjiangsu** [comment]
Do you want me to push the main changes to develop so that it's up to date?

**Comment by MrAntonS** [comment]
Mention this in #50
